dali/core/mm/default_resources.cc

```diff
@@ -366,32 +406,28 @@ device_async_resource *GetDefaultDeviceResource(int device_id) {
 }
 
 template <typename Kind>
-mm::pool_resource_base<Kind> *GetPoolInterface(mm::memory_resource<Kind> *mr) {
-  while (mr) {
-    if (auto *pool = dynamic_cast<mm::pool_resource_base<Kind>*>(mr))
-      return pool;
-    if (auto *up = dynamic_cast<mm::with_upstream<Kind>*>(mr)) {
-      mr = up->upstream();
-    } else {
-      break;
+void ReleaseUnusedMemory(mm::memory_resource<Kind> *mr) {
+  if (auto *pool = dynamic_cast<mm::pool_resource_base<Kind>*>(mr)) {
+    pool->release_unused();
+  } else if (auto *up_rsrc = dynamic_cast<mm::with_upstream<Kind>*>(mr)) {
+    ReleaseUnusedMemory(up_rsrc->upstream());
+  } else if (auto *bin_rsrc = dynamic_cast<mm::binning_resource_base<Kind>*>(mr)) {
+    for (int bin = 0; bin < bin_rsrc->num_bins(); bin++) {
+      ReleaseUnusedMemory(bin_rsrc->resource(bin));
     }
```

Comment:
Should we fail in all other cases?

Reply:
No - just do nothing. Originally, we looked for a pool resource and if not found, ignored it. Now I fused the search and freeing of the memory because, theoretically, a binning resource could have multiple pools, all of which might need freeing.

Reply:
Is there a valid use of this function in our code-base when it can fail to do anything? I am worried it may be error prone that we call a catch-all release function that can accidently not mention/forsee one of the resources and leak it.

Reply:
Well, yes. If we don't use a pool (but a resource that directly allocates) or when a user externally provides an allocator via C++ API (which is also possible), then not doing anything is quite expected. Until recently this function wasn't even called for host memory (which wasn't really intentional) but if it had, it wouldn't do anything anyway, because we simply used malloc.